### DISCUSSION TITLE
Creating efficient queries with Bookshelf to only include the fields that are required.

### DISCUSSION THREAD

**Discussion by mudassar-islam** [strapi/strapi#7538] [Legacy Topics]
I have a large Collection Type for my Products which includes each and every bit of detail about the product along with many relations to Categories, FAQ Articles are more.

It is of course not secure and not efficient at all to fetch every single field on every request, therefore I created a GET Route, a Controller and a Service and my goal is to only retrieve the fields that are required.

```js
module.exports = {
    home: async () => {
        const results = await strapi.query('product').model.query(qb => {
            qb.limit(3);
            qb.orderBy('id', 'DESC');
            qb.select(['id', 'name', 'sub_title', 'version']);
        }).fetch();

        return results;
    }
};
```

This is my service so far, It works properly but for all of the components / relations are visible. And I cannot seem to ignore them.
I want to ignore certain components / relations entirely, and others only specific fields.

For example I have a Media component which includes the cover and thumbnail, but I only want the thumbnail. Similarly I have a relation with FAQ Articles where I only wanna fetch the subject and slug.

I could map through the array and extract the fields, but I also don't want the Database to be queried for efficiency.
How can I achieve this ? Any help will be appreciated :)

**Comment by mudassar-islam** [comment]
Is there no way to do this or something :/ 
Having a hard time here..

I've managed to get Relations working in the way I want.
You just pass a `withRelated: []` property to your `fetch({})` call.
in that property you can define which relations you want to be included, and you can even pass query callbacks to select exactly what you want from the database. For example:
```js
module.exports = {
    home: async () => {
        const results = await strapi.query('product').model.query(qb => {
            qb.limit(3);
            qb.orderBy('id', 'DESC');
            qb.select(['id', 'name', 'sub_title', 'version']);
        }).fetch({
            withRelated: [
                'category'
            ]
        });

        return results;
    }
};
```

Im just looking for a solution for Components.

**Reply by derrickmehaffy** [reply]
Let me ping @petersg83 and @alexandrebodin to see if they have suggestions on this topic

**Reply by alexandrebodin** [reply]
There is currently no way to do this with booksheld as the auto population feature is implemented using the ORM lifecycles. This will be eventally refactored to allow for full control. 

What you can do however is use the underlying knex connection available in strapi.connections this will give you the full power of a SQL query builder without any ORM overhead.
